packages/tideline/test/tidelinedata_test.js

```diff
@@ -1009,4 +1009,58 @@ describe('TidelineData', function() {
     })
   });
 
+  describe('setEvents', function() {
+    var data = [
+      new types.DeviceEvent({
+        deviceTime: '2014-07-01T10:00:00',
+        eventId: 'Event1',
+        subType: 'zen',
+        inputTime: moment.utc().subtract(2, 'days').toISOString(),
+      }),
+      new types.DeviceEvent({
+        deviceTime: '2014-07-02T10:30:00',
+        inputTime: moment.utc().subtract(1, 'days').toISOString(),
+        eventId: 'Event2',
+        subType: 'zen',
+      }),
+      new types.DeviceEvent({
+        inputTime: moment.utc().subtract(4, 'hours').toISOString(),
+        eventId: 'Event3',
+        subType: 'zen',
+      }),
+      new types.DeviceEvent({
+        inputTime: moment.utc().subtract(3, 'hours').toISOString(),
+        eventId: 'Event3',
+        subType: 'zen',
+      }),
+    ];
+    var thisTd = new TidelineData(_.cloneDeep(data), {});
+
+    it('should be a function', function() {
+      assert.isFunction(thisTd.setEvents);
+    });
+
+    const Zen1 = _.filter(thisTd.zenEvents, { 'eventId': 'Event1' });
+    const Zen2 = _.filter(thisTd.zenEvents, { 'eventId': 'Event2' });
+    const Zen3 = _.filter(thisTd.zenEvents, { 'eventId': 'Event3' });
+
+    console.log(thisTd.zenEvents);
+    it('should deduplicate Events based on eventId', function() {
+      // PA1, PA2 and undefined
```

Comment:
Should we test a de-duplication ?

Reply:
That's what is done for Event3, 2 events as input and 1 event as output in zenEvents